11 add off gas rural clean

crispy-wonton opened the pull request:
Addresses #10 and #11 

# Description
Summary of changes:
- Add new get_data functions to load in supplementary data: off-gas postcodes oct 2023, and ONS postcode dir Aug 2023
- Join supplementary data into main MCS-EPC dataset

Note: the code now can take longer to run because of loading the ONSPD dataset. Total run time of script is approx 3-5 mins.

# Current caveats of supplementary data
**Off-gas status**
XOserve data ([see bottom of webpage for dataset](https://www.xoserve.com/help-centre/supply-points-metering/supply-point-administration-spa/)) seems to be main source of off-gas postcode information that I could find. I used the most recent dataset which is from October 2023, which contains active postcodes as of August 2023. The dataset only contains postcodes which are off-gas and no information about on-gas postcodes. Therefore when we perform the join with MCS-EPC data, we won’t know if postcodes where 'off-gas' is False are really *on* the gas grid or if the postcode's status is unknown by XOServe.

# Instructions for Reviewer
Please could you:
- Sense check the data used. As discussed, I have used 2011 RUC indicators from the [ONSPD Aug 2023](https://geoportal.statistics.gov.uk/datasets/487a5ba62c8b4da08f01eb3c08e304f6/about) which has 2011 data for England and Wales. I believe this uses RUC data for Scotland from 2011 as well even though there is 2020 data available, but I've kept as is because only 0.04% of data is missing RUC. I chose Aug 2023 ONSPD dataset as the current MCS-EPC data used in the analysis is 2023 Q2.
- Double check that "rural_urban_classification_mapping" in config is correct. I used the information on p19, section 39 in the [Aug 2023 ONSPD User Guide](https://www.ons.gov.uk/file?uri=/aboutus/transparencyandgovernance/freedomofinformationfoi/listofpostcodesandcorrespondinglocalauthorities/onspduserguideaug2023.pdf) for England/Wales codes and Table 2.3 of [Scottish Government Urban Rural Classification 2020](https://www.gov.scot/publications/scottish-government-urban-rural-classification-2020/pages/2/) for the Scotland codes.

To run the code:
- `python asf_heat_pump_affordability/pipeline/produce_costs_dataset.py --mcs_epc_join_date 231009 --cost_year_min 2021 --cpi_data_year 2023 --save_to_s3` - please could you test it still works for you as I've changed the getters. As usual please change the flags however you want.
- you can see the new columns added in the sample file that saves to S3 bucket. Or you can add `sample` to the return statement of `main()` and run `main()` in a notebook ¯\_(ツ)_/¯

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

crispy-wonton committed:
add off-gas and ruc data sources and ruc mapping to config

crispy-wonton committed:
add getters to get off-gas postcode dataset and ONS postcode directory

crispy-wonton committed:
add function to join supplementary data to MCS-EPC dataset
- join off gas status and rural-urban classification to MCS-EPC data
- add join function into main run script

crispy-wonton committed:
add json schema for loading ONSPD Aug 2023 file

crispy-wonton committed:
split get_df_from_url into multiple functions for different file types
update _ruc_code_conversion code to use string dtype only and no floats (floats no longer present due to addition of schema for ONSPD data)

crispy-wonton committed:
update getter function in run script

crispy-wonton commented on asf_heat_pump_affordability/getters/get_data.py:
```diff
@@ -1,20 +1,133 @@
+    return df
+
+
+def get_df_from_excel_url(url: str, **kwargs) -> pd.DataFrame:
```
I took a leaf from pandas here and separated the getters into different functions for different file types to avoid confusion/errors.

crispy-wonton commented on asf_heat_pump_affordability/getters/get_data.py:
```diff
@@ -1,20 +1,133 @@
+
+    Args
+        url (str): URL location of Excel file download
+        **kwargs for pandas.read_excel()
```
I'm not sure of the correct way to refer to kwargs from another function in docstrings and I was unable to find an example or practice to follow. If there's a better way please let me know :)

danlewis85 replied:
Basically I think this is fine, here's a couple of example of this off the top of my head: [seaborn.kdeplot](https://seaborn.pydata.org/generated/seaborn.kdeplot.html), [pandas.read_parquet](https://pandas.pydata.org/docs/reference/api/pandas.read_parquet.html). These look similar to what you've done, I don't think there is a definite approach.

crispy-wonton committed:
correct type hinting in _ruc_code_conversion function

crispy-wonton committed:
add getters functions to get IMD income data for Scotland and England & Wales
add source urls to config file for new datasets

crispy-wonton committed:
join IMD income data for Scotland and England & Wales to MCS-EPC sample dataset

crispy-wonton committed:
correct docstrings for IMD income data getters functions

danlewis85 commented on asf_heat_pump_affordability/getters/get_data.py:
```diff
@@ -1,20 +1,100 @@
+        ruc_code
+    ):  # convert RUC codes that appear as float values
+        ruc_code = str(int(ruc_code))
+    try:
```
if you use the `get` method of the dictionary, rather than indexing, you won't have to use a `try... except...` because `get` returns `None` when the key doesn't exist. Replace with:

`ruc = config["rural_urban_classification_mapping"].get(ruc_code)`

crispy-wonton replied:
Thank you! I forgot about get!!

danlewis85 commented on asf_heat_pump_affordability/getters/get_data.py:
```diff
@@ -1,20 +1,100 @@
+    Returns
+        pd.DataFrame: postcode directory for Great Britain
+    """
+    df = get_df_from_url(
```
Why go to the trouble of loading the whole onspd when you're only using the `pcd` and `ru11ind` columns. Passing kwargs so that `pandas.read_csv()` gets a `use_cols` argument should make this a lot quicker.

crispy-wonton replied:
it did make it a lot quicker

danlewis85 reviewed (changes requested):
This is broadly fine. On the onspd front, I don't think we need the whole dataset, in fact I think we just want the rural urban classification bit of it. Can you therefore just load in the relevant columns, transform and pass them through to the mcs-epc dataset.

In addition, you can address the minor point about the dictionary `get` method.

danlewis85 commented:
> This is broadly fine. On the onspd front, I don't think we need the whole dataset, in fact I think we just want the rural urban classification bit of it. Can you therefore just load in the relevant columns, transform and pass them through to the mcs-epc dataset.
> 
> In addition, you can address the minor point about the dictionary `get` method.

Looks like you'll also need to keep the 2011 LSOA/Datazone for the dependency joining IMS/SIMD.

crispy-wonton committed:
make changes requested in PR review
- use `get` for ruc code conversion instead of try/except
- read only relevant cols from ONS postcode directory to increase code speed

crispy-wonton committed:
Merge pull request #15 from nestauk/12-add-imd-income-deciles

12 add imd income deciles
